addons/tls13-middlebox/bootstrap.js

```diff
@@ -0,0 +1,276 @@
+
+    let securityInfo = getFieldValue(channel, "securityInfo");
+
+    if (securityInfo instanceof Ci.nsITransportSecurityInfo) {
+      securityInfo.QueryInterface(Ci.nsITransportSecurityInfo);
+
+      // extract security state and error code by which we can identify the reasons the connection failed
+      result.securityState = getFieldValue(securityInfo, "securityState");
+      result.errorCode = getFieldValue(securityInfo, "errorCode");
+    }
+
+    if (securityInfo instanceof Ci.nsISSLStatusProvider) {
+      securityInfo.QueryInterface(Ci.nsISSLStatusProvider);
+      let sslStatus = getFieldValue(securityInfo, "SSLStatus");
+
+      if (sslStatus) {
+        sslStatus.QueryInterface(Ci.nsISSLStatus);
+
+        // in case cert verification failed, we need to extract the cert chain from failedCertChain attribute
+        // otherwise, we extract cert chain using certDB.asyncVerifyCertAtTime API
+        let chain = null;
+
+        if (getFieldValue(securityInfo, "failedCertChain")) {
+          chain = nsIX509CertListToArray(securityInfo.failedCertChain);
+        } else {
+          chain = await getCertChain(getFieldValue(sslStatus, "serverCert"), CERT_USAGE_SSL_SERVER);
+        }
+
+        // extracting sha256 fingerprint for the leaf cert in the chain
+        result.serverSha256Fingerprint = getFieldValue(chain[0], "sha256Fingerprint");
+
+        // check the root cert to see if it is builtin certificate
+        result.isBuiltInRoot = (chain !== null && chain.length > 0) ? getFieldValue(chain[chain.length - 1], "isBuiltInRoot") : null;
+
+        // record the tls version Firefox ended up negotiating
+        result.protocolVersion = getFieldValue(sslStatus, "protocolVersion");
+      }
+    }
+  } catch (ex) {
+    result.exception = ex.message;
+  }
+
+  return result;
+}
+
+function makeRequest(config) {
+  return new Promise((resolve, reject) => {
+    // put together the configuration and the info collected from the connection
+    function reportResult(event, xhr) {
+      getInfo(xhr).then(info => {
+        let output = Object.assign({"result": {"event": event, "responseCode": xhr.status}}, config);
+        output.result = Object.assign(output.result, info);
+        resolve(output);
+
+        return true;
+      }).catch();
+    }
+
+    try {
+      // set the configuration to the values that were passed to this function
+      readwrite_prefs.set(VERSION_MAX_PREF, config.maxVersion);
+      readwrite_prefs.set(FALLBACK_LIMIT_PREF, config.fallbackLimit);
+
+      let xhr = Cc["@mozilla.org/xmlextras/xmlhttprequest;1"].createInstance(Ci.nsIXMLHttpRequest);
+
+      xhr.open("GET", `https://${config.website}`, true);
+
+      xhr.timeout = 10000;
+
+      xhr.channel.loadFlags |= Ci.nsIRequest.LOAD_ANONYMOUS;
+      xhr.channel.loadFlags |= Ci.nsIRequest.LOAD_BYPASS_CACHE;
+      xhr.channel.loadFlags |= Ci.nsIRequest.INHIBIT_CACHING;
+
+      xhr.addEventListener("load", e => {
```

Comment:
I *think* you've covered every condition below, but if you want to listen on something guaranteed to fire then you want a listener for `loadend` too.